src/query/api/v1/handler/influxdb/write.go

```diff
@@ -0,0 +1,285 @@
+	"github.com/m3db/m3/src/query/ts"
+	"github.com/m3db/m3/src/query/util/logging"
+
+	xerrors "github.com/m3db/m3/src/x/errors"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+	xtime "github.com/m3db/m3/src/x/time"
+	imodels "github.com/influxdata/influxdb/models"
+	"go.uber.org/zap"
+)
+
+const (
+	// InfluxWriteURL is the Influx DB write handler URL
+	InfluxWriteURL = handler.RoutePrefixV1 + "/influxdb/write"
+
+	// InfluxWriteHTTPMethod is the HTTP method used with this resource
+	InfluxWriteHTTPMethod = http.MethodPost
+)
+
+type ingestWriteHandler struct {
+	handlerOpts  options.HandlerOptions
+	promRewriter *promRewriter
+}
+
+type ingestField struct {
+	name  []byte // to be stored in __name__; rest of tags stay constant for the Point
+	value float64
+}
+
+type ingestIterator struct {
+	// what is being iterated
+	points       []imodels.Point
+	promRewriter *promRewriter
+	pointIndex   int
+	err          xerrors.MultiError
+
+	// following entries are within current point, and initialized
+	// when we go to the first entry in the current point
+	fields         []*ingestField
+	nextFieldIndex int
+	tags           models.Tags
+}
+
+func (self *ingestIterator) populateFields() bool {
+	point := self.points[self.pointIndex]
+	it := point.FieldIterator()
+	n := 0
+	self.fields = make([]*ingestField, 0, 10)
+	bname := make([]byte, 0, len(point.Name())+1)
+	bname = append(bname, point.Name()...)
+	bname = append(bname, byte('_'))
+	bnamelen := len(bname)
+	self.promRewriter.rewriteMetric(bname)
+	for it.Next() {
+		var value float64 = 0
+		n += 1
+		switch it.Type() {
+		case imodels.Boolean:
+			v, err := it.BooleanValue()
+			if err != nil {
+				self.err = self.err.Add(err)
+				continue
+			}
+			if v {
+				value = 1.0
+			}
+		case imodels.Integer:
+			v, err := it.IntegerValue()
+			if err != nil {
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = float64(v)
+		case imodels.Unsigned:
+			v, err := it.UnsignedValue()
+			if err != nil {
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = float64(v)
+		case imodels.Float:
+			v, err := it.FloatValue()
+			if err != nil {
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = v
+		default:
+			// TBD if we should stick strings as
+			// tags or not; to prevent cardinality
+			// explosion, we drop them for now
+			continue
+		}
+		tail := it.FieldKey()
+		name := make([]byte, 0, bnamelen+len(tail))
+		name = append(name, bname...)
+		name = append(name, tail...)
+		self.promRewriter.rewriteMetricTail(name[bnamelen:])
+		self.fields = append(self.fields, &ingestField{name: name, value: value})
+	}
+	return n > 0
+}
+
+func (self *ingestIterator) Next() bool {
+	for len(self.points) > self.pointIndex {
+		if self.nextFieldIndex == 0 {
+			// Populate tags only if we have fields we care about
+			if self.populateFields() {
+				point := self.points[self.pointIndex]
+				ptags := point.Tags()
+				tags := models.NewTags(len(ptags), nil)
```

Comment:
You'll want to flow the tag options from the server into the second argument to `models.NewTags(...)` here.  Otherwise the Tag ID scheme won't be the same as what is configured by the m3coordinator config.

See `tagOpts` and `tagOptions` in the Prometheus remote write endpoint for reference:
https://github.com/m3db/m3/blob/master/src/query/api/v1/handler/prometheus/remote/write.go